[GuardChecks] Cannot read _routeConfig of null

_From @francois-appliware on June 14, 2016 13:47_

Hi,

I am working with the new router and had an issue that I "managed to resolve", but as I don't quite understand how everything works under the hood I prefer to open this issue and try to understand what is going on.

In my application, I have a few routes as following :
```javascript
{ path: '/dashboard', component: DashboardComponent },
{
    path: '/companies', component: CompanyComponent, children: [
        { path: '/details/:id', component: CompanyDetailComponent },
        { path: '/new', component: CompanyEditComponent },
        { path: '/', component: CompanyDetailComponent, index: true }
    ]
},
{
    path: '/keys', component: KeyComponent, children: [
        { path: '/details/:id', component: KeyDetailComponent },
        { path: '/new', component: KeyEditComponent },
        { path: '/', component: CompanyDetailComponent, index: true }
    ]
}
```

When I am navigating from companies to keys, I have the following error : 

> Error {rejection: TypeError: Cannot read property '_routeConfig' of null
    at GuardChecks.runCanDeactivate (http://l…, promise: ZoneAwarePromise, zone: Zone, task: ZoneTask, message: "Uncaught (in promise): TypeError: Cannot read property '_routeConfig' of null"}

So I went ahead and downloaded the code to see what was going on inside that "GuardChecks".

For what I understand, it is used for the CanActivate / CanDeactivate (that I do not use in any of my components). Also, from what I understand, it is working with nullable values (in my case, where I don't have such guards).

Problem is, at some point in adds a check **with** a null value that throws the previous error mentioned above.

I managed to solve my issue by replacing the following line inside "router.js" inside the method _GuardChecks.prototype.runCanDeactivate = function (component, curr)_
`var canDeactivate = curr._routeConfig ? curr._routeConfig.canDeactivate : null;`
by the following : 
`var canDeactivate = curr ? (curr._routeConfig ? curr._routeConfig.canDeactivate : null) : null;`

(Yes, it's ugly but that's a hotfix so I can continue working)
I think the check can be set before the 'runCanDeactivate', where the checks are added, maybe in the traverseRoutes method inside GuardChecks.

Is this a known bug or a real bug and needs a fix or is it intended ?

Thank you.

_Copied from original issue: angular/vladivostok#48_

_From @maxailloud on June 20, 2016 9:24_

I have the same issue, not using guards at all:

```
browser_adapter.ts:74 Error: Uncaught (in promise): TypeError: Cannot read property '_routeConfig' of null
    at resolvePromise (zone.js:538)
    at zone.js:515
    at ZoneDelegate.invoke (zone.js:323)
    at Object.onInvoke (ng_zone_impl.ts:64)
    at ZoneDelegate.invoke (zone.js:322)
    at Zone.run (zone.js:216)
    at zone.js:571
    at ZoneDelegate.invokeTask (zone.js:356)
    at Object.onInvokeTask (ng_zone_impl.ts:53)
    at ZoneDelegate.invokeTask (zone.js:355)
```

--- a/modules/@angular/router/src/directives/router_link_active.ts
+++ b/modules/@angular/router/src/directives/router_link_active.ts
@@ -24,7 +24,7 @@ export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit 
   private classes: string[] = [];
   private subscription: Subscription;
 
-  @Input() private routerLinkActiveOptions: RouterLinkActiveOptions = {exact: true};
+  @Input() private routerLinkActiveOptions: RouterLinkActiveOptions = {exact: false};
 
   /**
    * @internal

--- a/modules/@angular/router/src/router.ts
+++ b/modules/@angular/router/src/router.ts
@@ -460,7 +460,7 @@ class GuardChecks {
   }
 
   private runCanDeactivate(component: Object, curr: ActivatedRouteSnapshot): Observable<boolean> {
-    const canDeactivate = curr._routeConfig ? curr._routeConfig.canDeactivate : null;
+    const canDeactivate = curr && curr._routeConfig ? curr._routeConfig.canDeactivate : null;
     if (!canDeactivate || canDeactivate.length === 0) return of (true);
     return Observable.from(canDeactivate)
         .map(c => {

--- a/modules/@angular/router/src/router_state.ts
+++ b/modules/@angular/router/src/router_state.ts
@@ -13,7 +13,7 @@ import {Observable} from 'rxjs/Observable';
 import {Route} from './config';
 import {PRIMARY_OUTLET, Params} from './shared';
 import {UrlPathWithParams, UrlSegment, UrlTree} from './url_tree';
-import {shallowEqual} from './utils/collection';
+import {shallowEqual, shallowEqualArrays} from './utils/collection';
 import {Tree, TreeNode} from './utils/tree';
 
 
@@ -188,10 +188,14 @@ function serializeNode(node: TreeNode<ActivatedRouteSnapshot>): string {
  * And we detect that by checking if the snapshot field is set.
  */
 export function advanceActivatedRoute(route: ActivatedRoute): void {
-  if (route.snapshot && !shallowEqual(route.snapshot.params, route._futureSnapshot.params)) {
+  if (route.snapshot) {
+    if (!shallowEqual(route.snapshot.params, route._futureSnapshot.params)) {
+      (<any>route.params).next(route._futureSnapshot.params);
+    }
+    if (!shallowEqualArrays(route.snapshot.url, route._futureSnapshot.url)) {
+      (<any>route.url).next(route._futureSnapshot.url);
+    }
     route.snapshot = route._futureSnapshot;
-    (<any>route.url).next(route.snapshot.url);
-    (<any>route.params).next(route.snapshot.params);
   } else {
     route.snapshot = route._futureSnapshot;
   }

--- a/modules/@angular/router/src/utils/collection.ts
+++ b/modules/@angular/router/src/utils/collection.ts
@@ -6,6 +6,14 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+export function shallowEqualArrays(a:any[], b:any[]):boolean {
+  if (a.length !== b.length) return false;
+  for (let i = 0; i < a.length; ++i) {
+    if (!shallowEqual(a[i], b[i])) return false;
+  }
+  return true;
+}
+
 export function shallowEqual(a: {[x: string]: any}, b: {[x: string]: any}): boolean {
   const k1 = Object.keys(a);
   const k2 = Object.keys(b);

--- a/modules/@angular/router/src/utils/collection.ts
+++ b/modules/@angular/router/src/utils/collection.ts
@@ -6,7 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-export function shallowEqualArrays(a:any[], b:any[]):boolean {
+export function shallowEqualArrays(a: any[], b: any[]): boolean {
   if (a.length !== b.length) return false;
   for (let i = 0; i < a.length; ++i) {
     if (!shallowEqual(a[i], b[i])) return false;

--- a/modules/@angular/router/test/router.spec.ts
+++ b/modules/@angular/router/test/router.spec.ts
@@ -133,17 +133,21 @@ describe('Integration', () => {
            const fixture = tcb.createFakeAsync(RootCmp);
            advance(fixture);
 
-           router.resetConfig([{path: '**', component: SimpleCmp}]);
+           router.resetConfig([{path: '**', component: CollectParamsCmp}]);
 
            router.navigateByUrl('/one/two');
            advance(fixture);
+           const cmp = fixture.debugElement.children[1].componentInstance;
            expect(location.path()).toEqual('/one/two');
-           expect(fixture.debugElement.nativeElement).toHaveText('simple');
+           expect(fixture.debugElement.nativeElement).toHaveText('collect-params');
+
+           expect(cmp.recordedUrls()).toEqual(['one/two']);
 
            router.navigateByUrl('/three/four');
            advance(fixture);
            expect(location.path()).toEqual('/three/four');
-           expect(fixture.debugElement.nativeElement).toHaveText('simple');
+           expect(fixture.debugElement.nativeElement).toHaveText('collect-params');
+           expect(cmp.recordedUrls()).toEqual(['one/two', 'three/four']);
          })));
 
   it('should support secondary routes',
@@ -1023,6 +1027,21 @@ class LinkWithQueryParamsAndFragment {
 
 @Component({selector: 'simple-cmp', template: `simple`, directives: ROUTER_DIRECTIVES})
 class SimpleCmp {
+}
+
+@Component({selector: 'collect-params-cmp', template: `collect-params`, directives: ROUTER_DIRECTIVES})
+class CollectParamsCmp {
+  private params: any = [];
+  private urls: any = [];
+
+  constructor(a: ActivatedRoute) {
+    a.params.forEach(p => this.params.push(p));
+    a.url.forEach(u => this.urls.push(u));
+  }
+
+  recordedUrls(): string[] {
+    return this.urls.map(a => a.map(p => p.path).join('/'));
+  }
 }
 
 @Component({selector: 'blank-cmp', template: ``, directives: ROUTER_DIRECTIVES})

--- a/modules/@angular/router/test/router.spec.ts
+++ b/modules/@angular/router/test/router.spec.ts
@@ -860,7 +860,7 @@ describe('Integration', () => {
   });
 
   describe('routerActiveLink', () => {
-    it('should set the class when the link is active (exact = true)',
+    it('should set the class when the link is active',
        fakeAsync(inject(
            [Router, TestComponentBuilder, Location],
            (router: Router, tcb: TestComponentBuilder, location: Location) => {
@@ -878,9 +878,9 @@ describe('Integration', () => {
                }]
              }]);
 
-             router.navigateByUrl('/team/22/link');
+             router.navigateByUrl('/team/22/link;exact=true');
              advance(fixture);
-             expect(location.path()).toEqual('/team/22/link');
+             expect(location.path()).toEqual('/team/22/link;exact=true');
 
              const native = fixture.debugElement.nativeElement.querySelector('a');
              expect(native.className).toEqual('active');
@@ -891,7 +891,7 @@ describe('Integration', () => {
              expect(native.className).toEqual('');
            })));
 
-    it('should set the class on a parent element when the link is active (exact = true)',
+    it('should set the class on a parent element when the link is active',
        fakeAsync(inject(
            [Router, TestComponentBuilder, Location],
            (router: Router, tcb: TestComponentBuilder, location: Location) => {
@@ -909,9 +909,9 @@ describe('Integration', () => {
                }]
              }]);
 
-             router.navigateByUrl('/team/22/link');
+             router.navigateByUrl('/team/22/link;exact=true');
              advance(fixture);
-             expect(location.path()).toEqual('/team/22/link');
+             expect(location.path()).toEqual('/team/22/link;exact=true');
 
              const native = fixture.debugElement.nativeElement.querySelector('link-parent');
              expect(native.className).toEqual('active');
@@ -922,7 +922,7 @@ describe('Integration', () => {
              expect(native.className).toEqual('');
            })));
 
-    it('should set the class when the link is active (exact = false)',
+    it('should set the class when the link is active',
        fakeAsync(inject(
            [Router, TestComponentBuilder, Location],
            (router: Router, tcb: TestComponentBuilder, location: Location) => {
@@ -940,9 +940,9 @@ describe('Integration', () => {
                }]
              }]);
 
-             router.navigateByUrl('/team/22/link;exact=false');
+             router.navigateByUrl('/team/22/link');
              advance(fixture);
-             expect(location.path()).toEqual('/team/22/link;exact=false');
+             expect(location.path()).toEqual('/team/22/link');
 
              const native = fixture.debugElement.nativeElement.querySelector('a');
              expect(native.className).toEqual('active');
@@ -988,18 +988,21 @@ class AbsoluteLinkCmp {
 class DummyLinkCmp {
   private exact: boolean;
   constructor(route: ActivatedRoute) {
-    // convert 'false' into false
-    this.exact = (<any>route.snapshot.params).exact !== 'false';
+    this.exact = (<any>route.snapshot.params).exact === 'true';
   }
 }
 
 @Component({
   selector: 'link-cmp',
   template:
-      `<router-outlet></router-outlet><link-parent routerLinkActive="active"><a [routerLink]="['./']">link</a></link-parent>`,
+      `<router-outlet></router-outlet><link-parent routerLinkActive="active" [routerLinkActiveOptions]="{exact: exact}"><a [routerLink]="['./']">link</a></link-parent>`,
   directives: ROUTER_DIRECTIVES
 })
 class DummyLinkWithParentCmp {
+  private exact: boolean;
+  constructor(route: ActivatedRoute) {
+    this.exact = (<any>route.snapshot.params).exact === 'true';
+  }
 }
 
 @Component({

--- a/modules/@angular/router/test/router.spec.ts
+++ b/modules/@angular/router/test/router.spec.ts
@@ -991,9 +991,7 @@ class AbsoluteLinkCmp {
 })
 class DummyLinkCmp {
   private exact: boolean;
-  constructor(route: ActivatedRoute) {
-    this.exact = (<any>route.snapshot.params).exact === 'true';
-  }
+  constructor(route: ActivatedRoute) { this.exact = (<any>route.snapshot.params).exact === 'true'; }
 }
 
 @Component({
@@ -1004,9 +1002,7 @@ class DummyLinkCmp {
 })
 class DummyLinkWithParentCmp {
   private exact: boolean;
-  constructor(route: ActivatedRoute) {
-    this.exact = (<any>route.snapshot.params).exact === 'true';
-  }
+  constructor(route: ActivatedRoute) { this.exact = (<any>route.snapshot.params).exact === 'true'; }
 }
 
 @Component({
@@ -1029,7 +1025,8 @@ class LinkWithQueryParamsAndFragment {
 class SimpleCmp {
 }
 
-@Component({selector: 'collect-params-cmp', template: `collect-params`, directives: ROUTER_DIRECTIVES})
+@Component(
+    {selector: 'collect-params-cmp', template: `collect-params`, directives: ROUTER_DIRECTIVES})
 class CollectParamsCmp {
   private params: any = [];
   private urls: any = [];
@@ -1039,9 +1036,7 @@ class CollectParamsCmp {
     a.url.forEach(u => this.urls.push(u));
   }
 
-  recordedUrls(): string[] {
-    return this.urls.map(a => a.map(p => p.path).join('/'));
-  }
+  recordedUrls(): string[] { return this.urls.map(a => a.map(p => p.path).join('/')); }
 }
 
 @Component({selector: 'blank-cmp', template: ``, directives: ROUTER_DIRECTIVES})

--- a/modules/@angular/router/test/url_serializer.spec.ts
+++ b/modules/@angular/router/test/url_serializer.spec.ts
@@ -101,6 +101,11 @@ describe('url serializer', () => {
     expect(tree.queryParams).toEqual({a: '(11)', b: '(22)'});
   });
 
+  it('should parse query params when with slashes', () => {
+    const tree = url.parse('/one?a=1/2&b=3/4');
+    expect(tree.queryParams).toEqual({a: '1/2', b: '3/4'});
+  });
+
   it('should parse key only query params', () => {
     const tree = url.parse('/one?a');
     expect(tree.queryParams).toEqual({a: 'true'});

